pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +59,91 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+// LazyLoadedHeadersSnapshotConfig stores information needed to track lazy loaded index headers.
+type LazyLoadedHeadersSnapshotConfig struct {
+	Path   string
+	UserID string
+}
+
+type lazyLoadedHeadersSnapshot struct {
+	// HeaderLastUsedTime is map of index header ulid.ULID to timestamp in millisecond.
+	HeaderLastUsedTime map[ulid.ULID]int64 `json:"header_last_used_time"`
+	UserID             string              `json:"user_id"`
+}
+
+// persist creates a file to store lazy loaded index header in atomic way.
+func (l lazyLoadedHeadersSnapshot) persist(persistPath string) error {
```

Comment:
Perhaps this would be clearer? In particular, would be good to be clear that the path passed in is a path to a directory, not a path to a file.

```suggestion
// persist atomically writes this snapshot to dir.
func (l lazyLoadedHeadersSnapshot) persist(dir string) error {
```